auth-foundation/src/test/java/com/okta/authfoundation/client/NetworkUtilsTest.kt

```diff
@@ -694,61 +695,158 @@ class NetworkingTest {
+
+        for (i in 0 until maxRetries + 1) { // One try + maxRetries retries
+            oktaRule.enqueue(path("/test")) { response ->
+                response.setResponseCode(429)
+                setRateLimitHeaders(response)
+            }
+        }
+
+        val request = Request.Builder()
+            .url(oktaRule.baseUrl.newBuilder().addPathSegments("test").build())
+            .build()
+        oktaRule.createOidcClient().performRequest(JsonObject.serializer(), request) {
+            it["foo"]!!.jsonPrimitive.content
+        }
+
+        assertThat(events.size).isEqualTo(11)
+        events.mapIndexed { index, event ->
+            assertThat(event.retryCount).isEqualTo(index)
+        }
+    }
+
+    @Test fun testPerformRequestRetryChangesDelayBasedOnHandledEvent(): Unit = runTest {
+        val exponentBase = 2.0
+        val events = mutableListOf<RateLimitExceededEvent>()
+        val rateLimitEventHandler = object : EventHandler {
+            override fun onEvent(event: Any) {
+                val errorEvent = event as RateLimitExceededEvent
+                events.add(event)
+                errorEvent.minDelaySeconds = exponentBase.pow(errorEvent.retryCount).toLong()
+            }
+        }
+        oktaRule.eventHandler.registerEventHandler(rateLimitEventHandler)
+
+        for (i in 0 until 4) {
+            oktaRule.enqueue(path("/test")) { response ->
+                response.setResponseCode(429)
+                setRateLimitHeaders(
+                    response,
+                    rateLimitResetEpochTime = TIME_EPOCH_SECS - 500,
+                )
+            }
+        }
+
+        val request = Request.Builder()
+            .url(oktaRule.baseUrl.newBuilder().addPathSegments("test").build())
+            .build()
+
+        val startTime = currentTime
+        oktaRule.createOidcClient().performRequest(JsonObject.serializer(), request) {
+            it["foo"]!!.jsonPrimitive.content
+        }
         val elapsedTime = currentTime - startTime
 
-        assertThat((result as OidcClientResult.Success<String>).result).isEqualTo("bar")
-        assertThat(elapsedTime).isEqualTo(minDelaySeconds.seconds.inWholeMilliseconds)
+        assertThat(events.size).isEqualTo(4)
+        assertThat(elapsedTime).isEqualTo((1 + 2 + 4).seconds.inWholeMilliseconds)
+    }
+
+    @Test fun testPerformRequestRetryChangesMaxRetriesBasedOnHandledEvent(): Unit = runTest {
+        val events = mutableListOf<RateLimitExceededEvent>()
+        val rateLimitEventHandler = object : EventHandler {
+            override fun onEvent(event: Any) {
+                val errorEvent = event as RateLimitExceededEvent
+                events.add(event)
+                errorEvent.maxRetries = if (errorEvent.retryCount == 10) 10 else 1000
+            }
+        }
+        oktaRule.eventHandler.registerEventHandler(rateLimitEventHandler)
+
+        for (i in 0 until 11) {
```

Comment:
Also might be worth a comment here explaining why 11. Similar to the above comment.